llm/gguf.go

```diff
@@ -199,21 +209,84 @@ func (llm *ggufModel) Decode(r io.Reader) error {
 
 	// decode tensors
 	for i := 0; uint64(i) < llm.NumTensor(); i++ {
-		if _, err := llm.readString(r); err != nil {
+		name, err := llm.readString(ro)
+		if err != nil {
 			return err
 		}
 
-		dimensions := llm.readU32(r)
+		dims := llm.readU32(ro)
+
+		nes := [4]uint64{1, 1, 1, 1}
+		for i := 0; uint32(i) < dims; i++ {
+			nes[i] = llm.readU64(ro)
+		}
+
+		kind := llm.readU32(ro)
+		offset := llm.readU64(ro)
+
+		var blockSize uint64
+		switch {
+		case kind < 2:
```

Comment:
What do these kind numbers correspond to?

Reply:
this is the quantization type, FP32, FP16, Q4_0, Q5_0, etc. of each layer. it's not quite the same as the file type since that is the aggregate type of the model